critwall/src/main/java/org/opencadc/critwall/ThreadPool.java

```diff
@@ -0,0 +1,179 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
+import org.apache.log4j.Logger;
+
+/**
+ * Implementation of a thread pool executor.
+ * Thread pool is fixed size. At startup, threads are set to consume the contents of the BlockingQueue.
+ * Threads in the pool are blocked when trying to pull tasks from an empty queue. (see API for BlockingQueue.take().)
+ */
+public class ThreadPool {
+    private static Logger log = Logger.getLogger(ThreadPool.class);
+
+    private final String poolBasename = ThreadPool.class.getName();
+    private final BlockingQueue<Runnable> taskQueue;
+    private final int nthreads;
+    private final ArrayList threads;
+
+
+    public ThreadPool(BlockingQueue<Runnable> blockingQueue, int nthreads) {
+
+        if (nthreads <= 0) {
+            String errMsg = "nthreads should > 1 (" + nthreads + ")";
+            log.error(errMsg);
+            throw new InvalidParameterException(errMsg);
+        }
+
+        this.taskQueue = blockingQueue;
+        this.nthreads = nthreads;
+        this.threads = new ArrayList(nthreads);
+
+        log.info(poolBasename + " - starting up");
+        log.debug("initial thread count: " + threads.size() + " requested size: " + nthreads);
+
+        synchronized (threads) {
+            while (threads.size() < nthreads) {
+                int threadNum = threads.size() + 1;
+                log.debug("adding worker thread " + threadNum);
+                WorkerThread t = new WorkerThread(threadNum);
+                t.setPriority(Thread.MIN_PRIORITY);
+                threads.add(t);
+                t.start();
+            }
+        }
+        log.debug("after pool startup - thread count: " + threads.size() + " requested size: " + nthreads);
+        log.debug(poolBasename + " - ctor done");
+    }
+
+    public void terminate() {
+        log.debug(poolBasename + ".terminate() starting");
+
+        // terminate thread pool members
+        synchronized (threads) {
+            Iterator<WorkerThread> threadIter = threads.iterator();
+            while (threadIter.hasNext()) {
+                WorkerThread t = threadIter.next();
+                log.debug(poolBasename + ".terminate() interrupting WorkerThread " + t.getName());
+                synchronized (t) {
+                    log.debug(poolBasename + ".terminate(): interrupting " + t.getName());
+                    t.interrupt();
+                }
+            }
+        }
+
+        log.debug(poolBasename + ".terminate() DONE");
+    }
+
+    private class WorkerThread extends Thread {
+        Runnable currentTask;
+
+        WorkerThread(int threadNum) {
+            super();
+            setDaemon(true);
+            setName(poolBasename + "-" + threadNum);
+        }
+
+        // threads keep running as long as they are in the threads list
+        public void run() {
+            log.debug(poolBasename + " - START");
+            boolean cont = true;
```

Comment:
This loop and terminate() no longer work as they did in DLM, because in DLM the pool was trying to gracefully change size without losing any tasks. Here, an `interrupt()` is only for causing the thread to exit (run() to return) more or less immediately. In that case, this needs the much simpler:
```
while (true) {
    try {
          Runnable task = taskQueue.take();
          task.run();
    } catch (InterruptedException ex) {
           return;
    } catch (Exception ignore) {
            log.debug("poorly behaved task threw an exception", ignore);
    }
}
```
Basically, if you look at this method more closely you realise that most of the lines of code no longer do anything useful :-)

The class javadoc for ThreadPool should note that some jobs in the queue may be lost if the pool is terminated while working.

Reply:
done (and done, for the javadoc update)